Kafka stream state stores cannot be found after upgrade to 3.2.5

**Describe the issue**
The following code works with SCS 3.2.4:
```kotlin
@Service
class TechnischerplatzAdresseHandler(private val interactiveQueryService: InteractiveQueryService) {
    private val technischerplatzAdresseStore: ReadOnlyKeyValueStore<String, TechnischerplatzAdresse>
        get() = interactiveQueryService.getStore("technischerplatzadresse-store")

    suspend fun get(request: ServerRequest) = withContext(Dispatchers.IO) {
        technischerplatzAdresseStore["1234"]?.let {
            ServerResponse.ok().build().awaitSingle()
        } ?: ServerResponse.notFound().buildAndAwait()
    }
}
```

But it raises an exception with SCS 3.2.5:
```
Caused by: org.apache.kafka.streams.errors.UnknownStateStoreException: Cannot get state store technischerplatzadresse-store because no such store is registered in the topology.
	at org.apache.kafka.streams.KafkaStreams.store(KafkaStreams.java:1603)
	at org.springframework.cloud.stream.binder.kafka.streams.InteractiveQueryService.getStateStoreFromKafkaStreams(InteractiveQueryService.java:121)
	at org.springframework.cloud.stream.binder.kafka.streams.InteractiveQueryService.lambda$getQueryableStore$0(InteractiveQueryService.java:109)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
	at org.springframework.cloud.stream.binder.kafka.streams.InteractiveQueryService.getQueryableStore(InteractiveQueryService.java:88)
	at at.wrwks.bef.projekt.domain.technischerplatzadresse.TechnischerplatzAdresseHandler$get$2.invokeSuspend(TechnischerplatzAdresseHandler.kt:39)
```

**To Reproduce**
Try to load a state store as described

**Version of the framework**
Spring Cloud ~~2020.0.4~~ 2021.0.4 with Spring Cloud Stream 3.2.5

**Expected behavior**
**Screenshots**

**Additional context**
With debugging I got down to `StreamsMetadataState#isInitialized()` where `clusterMetadata.topics().isEmpty() == true` which it shouldn't be. It is located in `kafka-streams:3.1.2` on line 321.
The topology hasn't changed and the store is there.

--- a/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/InteractiveQueryService.java
+++ b/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/InteractiveQueryService.java
@@ -16,6 +16,8 @@
 
 package org.springframework.cloud.stream.binder.kafka.streams;
 
+import java.lang.reflect.Field;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
@@ -32,6 +34,7 @@ import org.apache.kafka.streams.StoreQueryParameters;
 import org.apache.kafka.streams.StreamsConfig;
 import org.apache.kafka.streams.errors.InvalidStateStoreException;
 import org.apache.kafka.streams.errors.UnknownStateStoreException;
+import org.apache.kafka.streams.processor.internals.TopologyMetadata;
 import org.apache.kafka.streams.state.HostInfo;
 import org.apache.kafka.streams.state.QueryableStoreType;
 import org.apache.kafka.streams.state.StreamsMetadata;
@@ -41,6 +44,7 @@ import org.springframework.retry.RetryPolicy;
 import org.springframework.retry.backoff.FixedBackOffPolicy;
 import org.springframework.retry.policy.SimpleRetryPolicy;
 import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.ReflectionUtils;
 import org.springframework.util.StringUtils;
 
 /**
@@ -53,6 +57,7 @@ import org.springframework.util.StringUtils;
  * @author Renwei Han
  * @author Serhii Siryi
  * @author Nico Pommerening
+ * @author Chris Bono
  * @since 2.1.0
  */
 public class InteractiveQueryService {
@@ -63,6 +68,8 @@ public class InteractiveQueryService {
 
 	private final KafkaStreamsBinderConfigurationProperties binderConfigurationProperties;
 
+	private final Field topologyMetadataField;
+
 	/**
 	 * Constructor for InteractiveQueryService.
 	 * @param kafkaStreamsRegistry holding {@link KafkaStreamsRegistry}
@@ -72,6 +79,8 @@ public class InteractiveQueryService {
 			KafkaStreamsBinderConfigurationProperties binderConfigurationProperties) {
 		this.kafkaStreamsRegistry = kafkaStreamsRegistry;
 		this.binderConfigurationProperties = binderConfigurationProperties;
+		this.topologyMetadataField = ReflectionUtils.findField(KafkaStreams.class, "topologyMetadata");
+		this.topologyMetadataField.setAccessible(true);
 	}
 
 	/**
@@ -85,15 +94,16 @@ public class InteractiveQueryService {
 
 		KafkaStreams contextSpecificKafkaStreams = getThreadContextSpecificKafkaStreams();
 
+		final StoreQueryParameters<T> storeQueryParams = StoreQueryParameters.fromNameAndType(storeName, storeType);
+
 		return getRetryTemplate().execute(context -> {
 			T store = null;
 			Throwable throwable = null;
 			if (contextSpecificKafkaStreams != null) {
 				try {
-					store = contextSpecificKafkaStreams.store(StoreQueryParameters.fromNameAndType(storeName, storeType));
+					store = contextSpecificKafkaStreams.store(storeQueryParams);
 				}
 				catch (InvalidStateStoreException e) {
-					// pass through..
 					throwable = e;
 				}
 			}
@@ -104,28 +114,51 @@ public class InteractiveQueryService {
 				LOG.warn("Store (" + storeName + ") could not be found in Streams context, falling back to all known Streams instances");
 			}
 
+			// Find all apps that know about the store
+			Map<KafkaStreams, T> candidateStores = new HashMap<>();
 			for (KafkaStreams kafkaStreamApp : kafkaStreamsRegistry.getKafkaStreams()) {
 				try {
-					return getStateStoreFromKafkaStreams(kafkaStreamApp, storeName, storeType);
+					candidateStores.put(kafkaStreamApp, kafkaStreamApp.store(storeQueryParams));
 				}
 				catch (Exception ex) {
 					throwable = ex;
 				}
 			}
+
+			// Store exists in a single app - no further resolution required
+			if (candidateStores.size() == 1) {
+				return candidateStores.values().stream().findFirst().get();
+			}
+
+			// If the store is in multiple streams apps - discard any apps that do not actually have the store
+			if (candidateStores.size() > 1) {
+
+				candidateStores = candidateStores.entrySet().stream()
+						.filter((e) -> storeActuallyAvailable(e.getKey(), storeName))
+						.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+				if (candidateStores.size() == 1) {
+					return candidateStores.values().stream().findFirst().get();
+				}
+
+				throwable = (candidateStores.size() == 0) ?
+						new UnknownStateStoreException("Store (" + storeName + ") not available to Streams instance") :
+						new InvalidStateStoreException("Store (" + storeName + ") available to more than one Streams instance");
+
+			}
 			throw new IllegalStateException("Error retrieving state store: " + storeName, throwable);
 		});
 	}
 
-	private <T> T getStateStoreFromKafkaStreams(KafkaStreams kafkaStreams, String storeName, QueryableStoreType<T> storeType) {
-		// Check KafkaStreams app knows about the state store
-		T store = kafkaStreams.store(StoreQueryParameters.fromNameAndType(storeName, storeType));
-
-		// Check KafkaStreams app actually has the state store
-		if (kafkaStreams.streamsMetadataForStore(storeName).stream()
-				.noneMatch((sm) -> sm.stateStoreNames().contains(storeName))) {
-			throw new UnknownStateStoreException("Store (" + storeName + ") not available to Streams instance");
+	private boolean storeActuallyAvailable(KafkaStreams kafkaStreams, String storeName) {
+		try {
+			TopologyMetadata topologyMetadata = (TopologyMetadata) ReflectionUtils.getField(topologyMetadataField, kafkaStreams);
+			return !topologyMetadata.sourceTopicsForStore(storeName, null).isEmpty();
 		}
-		return store;
+		catch (Exception ex) {
+			LOG.error("Unable to determine if store (" + storeName + ") is available to the app due to: " + ex.getMessage(), ex);
+		}
+		return false;
 	}
 
 	/**

--- a/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/InteractiveQueryServiceMultiStateStoreTests.java
+++ b/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/InteractiveQueryServiceMultiStateStoreTests.java
@@ -63,7 +63,19 @@ class InteractiveQueryServiceMultiStateStoreTests {
 	private static final EmbeddedKafkaBroker embeddedKafka = EmbeddedKafkaCondition.getBroker();
 
 	@Test
-	void stateStoreOnlyAvailableOnKafkaStreamsAppWhereItIsUsed() {
+	void stateStoreAvailableOnProperAppWhenAppServerPropertySet() {
+		stateStoreAvailableOnProperApp(true);
+	}
+
+	@Test
+	void stateStoreAvailableOnProperAppWhenAppServerPropertyNotSet() {
+		stateStoreAvailableOnProperApp(false);
+	}
+
+	private void stateStoreAvailableOnProperApp(boolean shouldSetAppServerProperty) {
+		String appServerArg = shouldSetAppServerProperty ?
+				"--spring.cloud.stream.kafka.streams.binder.configuration.application.server=" + embeddedKafka.getBrokersAsString() :
+				"--foo=bar";
 		try (ConfigurableApplicationContext context = new SpringApplicationBuilder()
 				.sources(MultipleAppsWithUsedStateStoresTestApplication.class)
 				.web(WebApplicationType.NONE)
@@ -74,16 +86,16 @@ class InteractiveQueryServiceMultiStateStoreTests {
 						"--spring.cloud.stream.function.bindings.app2-in-0=input2",
 						"--spring.cloud.stream.kafka.streams.binder.functions.app1.application-id=stateStoreTestApp1",
 						"--spring.cloud.stream.kafka.streams.binder.functions.app2.application-id=stateStoreTestApp2",
-						"--spring.cloud.stream.kafka.streams.binder.configuration.application.server="
-								+ embeddedKafka.getBrokersAsString(),
-						"--spring.cloud.stream.kafka.streams.binder.brokers="
-								+ embeddedKafka.getBrokersAsString())
+						appServerArg,
+						"--spring.cloud.stream.kafka.streams.binder.brokers=" + embeddedKafka.getBrokersAsString())
 		) {
 			waitForRunningStreams(context.getBean(KafkaStreamsRegistry.class));
+
+			InteractiveQueryService queryService = context.getBean(InteractiveQueryService.class);
+
 			// The KafkaStreams.store() used by query service is non-deterministic so perform the operation multiple times to
 			// surface any possible issues. Also, no need to actually write anything to the stores, the store.get() call will
 			// cause a failure when the state store is invalid.
-			InteractiveQueryService queryService = context.getBean(InteractiveQueryService.class);
 			for (int i = 0; i < 100; i++) {
 				assertThat(queryService.getQueryableStore(STORE_1_NAME, QueryableStoreTypes.keyValueStore())
 						.get("someKey")).isNull();

--- a/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/KafkaStreamsInteractiveQueryIntegrationTests.java
+++ b/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/KafkaStreamsInteractiveQueryIntegrationTests.java
@@ -95,7 +95,7 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 	}
 
 	@Test
-	void testStateStoreRetrievalRetry() {
+	void stateStoreRetrievalRetriedOnFailure() {
 
 		StreamsBuilderFactoryBean mock = Mockito.mock(StreamsBuilderFactoryBean.class);
 		KafkaStreams mockKafkaStreams = Mockito.mock(KafkaStreams.class);
@@ -124,7 +124,7 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 	}
 
 	@Test
-	void testStateStoreRetrievalRetryForHostInfoService() {
+	void hostInfoRetrievalRetriedOnFailure() {
 		StreamsBuilderFactoryBean mock = Mockito.mock(StreamsBuilderFactoryBean.class);
 		KafkaStreams mockKafkaStreams = Mockito.mock(KafkaStreams.class);
 		Mockito.when(mock.getKafkaStreams()).thenReturn(mockKafkaStreams);
@@ -153,7 +153,7 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 	}
 
 	@Test
-	void testKstreamBinderWithPojoInputAndStringOuput() {
+	void queryServiceProvidesStateStoreAndHostInfo() {
 		SpringApplication app = new SpringApplication(ProductCountApplication.class);
 		app.setWebApplicationType(WebApplicationType.NONE);
 		ConfigurableApplicationContext context = app.run("--server.port=0",
@@ -167,7 +167,7 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 						+ "=org.apache.kafka.common.serialization.Serdes$StringSerde",
 				"--spring.cloud.stream.kafka.streams.binder.configuration.default.value.serde"
 						+ "=org.apache.kafka.common.serialization.Serdes$StringSerde",
-				"--spring.cloud.stream.kafka.streams.bindings.input.consumer.applicationId=ProductCountApplication-abc",
+				"--spring.cloud.stream.kafka.streams.bindings.input.consumer.applicationId=ProductCountApplication-def",
 				"--spring.cloud.stream.kafka.streams.binder.configuration.application.server"
 						+ "=" + embeddedKafka.getBrokersAsString(),
 				"--spring.cloud.stream.kafka.streams.binder.brokers="
@@ -187,13 +187,13 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf, true);
 		template.setDefaultTopic("foos");
 		template.sendDefault("{\"id\":\"123\"}");
-		ConsumerRecord<String, String> cr = KafkaTestUtils.getSingleRecord(consumer,
-				"counts-id");
-		assertThat(cr.value().contains("Count for product with ID 123: 1")).isTrue();
 
-		ProductCountApplication.Foo foo = context
-				.getBean(ProductCountApplication.Foo.class);
-		assertThat(foo.getProductStock(123).equals(1L));
+		ConsumerRecord<String, String> cr = KafkaTestUtils.getSingleRecord(consumer, "counts-id");
+		assertThat(cr.value()).isIn("Count for product with ID 123: 1", "Count for product with ID 123: 2");
+
+		// Make sure the count can be retrieved from local state store
+		ProductCountApplication.Foo foo = context.getBean(ProductCountApplication.Foo.class);
+		assertThat(foo.getProductStock(123)).isIn(1L, 2L);
 
 		// perform assertions on HostInfo related methods in InteractiveQueryService
 		InteractiveQueryService interactiveQueryService = context
@@ -232,6 +232,41 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 				.isEqualTo(embeddedKafka.getBrokersAsString());
 	}
 
+	@Test
+	void queryServiceProvidesStateStoreWithoutSettingAppServerProperty() {
+		SpringApplication app = new SpringApplication(ProductCountApplication.class);
+		app.setWebApplicationType(WebApplicationType.NONE);
+		try (ConfigurableApplicationContext context = app.run("--server.port=0",
+				"--spring.jmx.enabled=false",
+				"--spring.cloud.stream.function.bindings.process-in-0=input",
+				"--spring.cloud.stream.function.bindings.process-out-0=output",
+				"--spring.cloud.stream.bindings.input.destination=foos",
+				"--spring.cloud.stream.bindings.output.destination=counts-id",
+				"--spring.cloud.stream.kafka.streams.binder.configuration.commit.interval.ms=1000",
+				"--spring.cloud.stream.kafka.streams.binder.configuration.default.key.serde"
+						+ "=org.apache.kafka.common.serialization.Serdes$StringSerde",
+				"--spring.cloud.stream.kafka.streams.binder.configuration.default.value.serde"
+						+ "=org.apache.kafka.common.serialization.Serdes$StringSerde",
+				"--spring.cloud.stream.kafka.streams.bindings.input.consumer.applicationId=ProductCountApplication-def",
+				"--spring.cloud.stream.kafka.streams.binder.brokers=" + embeddedKafka.getBrokersAsString())
+		) {
+			// Send an input product record
+			Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+			DefaultKafkaProducerFactory<Integer, String> pf = new DefaultKafkaProducerFactory<>(senderProps);
+			KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf, true);
+			template.setDefaultTopic("foos");
+			template.sendDefault("{\"id\":\"123\"}");
+
+			// Wait for output from the streams app
+			ConsumerRecord<String, String> cr = KafkaTestUtils.getSingleRecord(consumer, "counts-id");
+			assertThat(cr.value()).isIn("Count for product with ID 123: 1", "Count for product with ID 123: 2");
+
+			// Make sure the count can be retrieved from local state store
+			ProductCountApplication.Foo foo = context.getBean(ProductCountApplication.Foo.class);
+			assertThat(foo.getProductStock(123)).isIn(1L, 2L);
+		}
+	}
+
 	@EnableAutoConfiguration
 	public static class ProductCountApplication {
 
@@ -243,8 +278,7 @@ public class KafkaStreamsInteractiveQueryIntegrationTests {
 					.groupByKey(Grouped.with(new Serdes.IntegerSerde(),
 							new JsonSerde<>(Product.class)))
 					.count(Materialized.as("prod-id-count-store")).toStream()
-					.map((key, value) -> new KeyValue<>(null,
-							"Count for product with ID 123: " + value));
+					.map((key, value) -> new KeyValue<>(null, "Count for product with ID 123: " + value));
 		}
 
 		@Bean

